File: android/app/src/main/java/com/personaltasks/voice/SettingsActivity.java
package com.personaltasks.voice;

import android.app.Activity;
import android.graphics.Typeface;
import android.os.Bundle;
import android.text.InputType;
import android.view.Gravity;
import android.view.ViewGroup;
import android.widget.Button;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.Switch;
import android.widget.TextView;
import android.widget.Toast;

import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;

public class SettingsActivity extends Activity {
    private EditText baseUrl;
    private EditText token;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        LinearLayout root = Ui.page(this);
        setContentView(Ui.scrollPage(this, root));

        LinearLayout header = Ui.row(this);
        TextView title = Ui.title(this, "Настройки");
        header.addView(title, Ui.matchWeight(1));
        TextView close = Ui.iconButton(this, "×");
        close.setContentDescription("Закрыть настройки");
        close.setOnClickListener(v -> finish());
        header.addView(close);
        root.addView(header, Ui.matchWrap());
        root.addView(Ui.spacer(this, 22));

        root.addView(generalCard(), Ui.matchWrap());
        root.addView(Ui.spacer(this, 18));
        root.addView(appearanceCard(), Ui.matchWrap());
        root.addView(Ui.spacer(this, 18));
        root.addView(connectionCard(), Ui.matchWrap());
        root.addView(Ui.spacer(this, 18));
        root.addView(aboutCard(), Ui.matchWrap());
        root.addView(Ui.spacer(this, 18));
        root.addView(developerCard(), Ui.matchWrap());
    }

    private LinearLayout generalCard() {
        LinearLayout card = Ui.card(this, 18);
        card.addView(Ui.section(this, "Общие"));

        Switch sound = switchRow("Звук", "Включить звуковые эффекты", AppPrefs.soundEnabled(this));
        Switch vibration = switchRow("Вибрация", "Короткая вибрация при ответе", AppPrefs.vibrationEnabled(this));
        sound.setOnCheckedChangeListener((buttonView, isChecked) -> AppPrefs.saveGeneral(this, isChecked, vibration.isChecked()));
        vibration.setOnCheckedChangeListener((buttonView, isChecked) -> AppPrefs.saveGeneral(this, sound.isChecked(), isChecked));
        card.addView(sound, Ui.matchWrap());
        card.addView(vibration, Ui.matchWrap());
        return card;
    }

    private LinearLayout appearanceCard() {
        LinearLayout card = Ui.card(this, 18);
        card.addView(Ui.section(this, "Внешний вид"));
        LinearLayout row = Ui.row(this);
        String active = AppPrefs.themeMode(this);
        row.addView(themeButton("Системная", "system", active), Ui.matchWeight(1));
        row.addView(themeButton("Светлая", "light", active), Ui.matchWeight(1));
        row.addView(themeButton("Тёмная", "dark", active), Ui.matchWeight(1));
        card.addView(row, Ui.matchWrap());
        Ui.margin(row, 0, 14, 0, 10);
        card.addView(Ui.subtitle(this, "Тема сохраняется локально. Полный dark-mode можно довести отдельным этапом."));
        return card;
    }

    private TextView themeButton(String label, String mode, String active) {
        boolean selected = mode.equals(active);
        TextView tv = Ui.text(this, selected ? label + " ✓" : label, 15, selected ? Ui.PRIMARY : Ui.TEXT, Typeface.BOLD);
        tv.setGravity(Gravity.CENTER);
        tv.setPadding(Ui.dp(this, 8), Ui.dp(this, 18), Ui.dp(this, 8), Ui.dp(this, 18));
        tv.setBackground(Ui.round(this, selected ? Ui.PRIMARY_SOFT : Ui.SURFACE_SOFT, Ui.dp(this, 18), selected ? Ui.PRIMARY : Ui.BORDER, 1));
        tv.setOnClickListener(v -> {
            AppPrefs.saveThemeMode(this, mode);
            recreate();
        });
        return tv;
    }

    private LinearLayout connectionCard() {
        LinearLayout card = Ui.card(this, 18);
        card.addView(Ui.section(this, "Подключение"));
        LinearLayout row = Ui.row(this);
        OrbView mic = new OrbView(this);
        row.addView(mic, Ui.lp(Ui.dp(this, 88), Ui.dp(this, 88)));

        LinearLayout info = new LinearLayout(this);
        info.setOrientation(LinearLayout.VERTICAL);
        boolean connected = !AppPrefs.deviceToken(this).isEmpty();
        info.addView(Ui.text(this, connected ? "Устройство готово к записи" : "Устройство не подключено", 18, Ui.TEXT, Typeface.BOLD));
        info.addView(Ui.chip(this, connected ? "Подключено" : "Не подключено", connected ? Ui.SUCCESS : Ui.DANGER, connected ? 0xFFD1FAE5 : 0xFFFFE4E6));
        info.addView(Ui.subtitle(this, "Последняя синхронизация: " + lastSyncText()));
        row.addView(info, Ui.matchWeight(1));
        Ui.margin(info, 16, 0, 0, 0);
        card.addView(row, Ui.matchWrap());
        return card;
    }

    private LinearLayout aboutCard() {
        LinearLayout card = Ui.card(this, 18);
        card.addView(Ui.section(this, "О приложении"));
        card.addView(Ui.text(this, "Голосовые задачи", 18, Ui.TEXT, Typeface.BOLD));
        card.addView(Ui.subtitle(this, "Версия 0.1.0\nТокен: " + AppPrefs.maskedToken(this)));
        return card;
    }

    private LinearLayout developerCard() {
        LinearLayout card = Ui.card(this, 18);
        card.addView(Ui.section(this, "Для разработчиков"));
        card.addView(Ui.subtitle(this, "Здесь можно заменить backend URL и mobile device token."));
        card.addView(Ui.spacer(this, 12));

        baseUrl = input("Backend URL");
        baseUrl.setText(AppPrefs.baseUrl(this));
        card.addView(baseUrl, Ui.matchWrap());
        Ui.margin(baseUrl, 0, 0, 0, 10);

        token = input("Mobile device token");
        token.setInputType(InputType.TYPE_CLASS_TEXT | InputType.TYPE_TEXT_VARIATION_VISIBLE_PASSWORD);
        token.setText(AppPrefs.deviceToken(this));
        card.addView(token, Ui.matchWrap());

        Button save = Ui.button(this, "Сохранить подключение", true);
        save.setOnClickListener(v -> {
            AppPrefs.save(this, baseUrl.getText().toString(), token.getText().toString());
            AppPrefs.markSynced(this);
            Toast.makeText(this, "Настройки сохранены", Toast.LENGTH_SHORT).show();
            recreate();
        });
        card.addView(save, Ui.matchWrap());
        Ui.margin(save, 0, 14, 0, 0);
        return card;
    }

    private Switch switchRow(String title, String description, boolean checked) {
        Switch sw = new Switch(this);
        sw.setText(title + "\n" + description);
        sw.setTextSize(16);
        sw.setTextColor(Ui.TEXT);
        sw.setChecked(checked);
        sw.setPadding(0, Ui.dp(this, 14), 0, Ui.dp(this, 14));
        return sw;
    }

    private EditText input(String hint) {
        EditText edit = new EditText(this);
        edit.setHint(hint);
        edit.setTextSize(16);
        edit.setSingleLine(true);
        edit.setMinHeight(Ui.dp(this, 56));
        edit.setPadding(Ui.dp(this, 14), 0, Ui.dp(this, 14), 0);
        edit.setBackground(Ui.round(this, Ui.SURFACE, Ui.dp(this, 16), Ui.BORDER, 1));
        return edit;
    }

    private String lastSyncText() {
        long value = AppPrefs.lastSyncAt(this);
        if (value <= 0) return "ещё не было";
        return new SimpleDateFormat("сегодня, HH:mm", Locale.getDefault()).format(new Date(value));
    }
}

File: android/app/src/main/java/com/personaltasks/voice/SettingsActivity.kt
package com.personaltasks.voice

import android.os.Bundle
import android.widget.Toast
import androidx.activity.ComponentActivity
import androidx.activity.compose.setContent
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Switch
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class SettingsActivity : ComponentActivity() {
    private var sound by mutableStateOf(true)
    private var vibration by mutableStateOf(true)
    private var theme by mutableStateOf("system")
    private var baseUrl by mutableStateOf("")
    private var token by mutableStateOf("")
    private var developerOpen by mutableStateOf(false)

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        load()
        setContent { SettingsScreen() }
    }

    private fun load() {
        sound = AppPrefs.soundEnabled(this)
        vibration = AppPrefs.vibrationEnabled(this)
        theme = AppPrefs.themeMode(this)
        baseUrl = AppPrefs.baseUrl(this)
        token = AppPrefs.deviceToken(this)
    }

    @Composable
    private fun SettingsScreen() {
        VoicePage(scroll = true) {
            Header("Настройки", "Голосовые задачи, подключение и внешний вид.", action = { RoundIcon("×", onClick = { finish() }) })
            Spacer(Modifier.height(24.dp))

            GlassCard {
                SectionLabel("Общие")
                SettingSwitch("Звук", "Включить звуковые эффекты", sound) {
                    sound = it
                    AppPrefs.saveGeneral(this@SettingsActivity, sound, vibration)
                }
                SettingSwitch("Вибрация", "Короткая вибрация при ответе", vibration) {
                    vibration = it
                    AppPrefs.saveGeneral(this@SettingsActivity, sound, vibration)
                }
            }
            Spacer(Modifier.height(18.dp))

            GlassCard {
                SectionLabel("Внешний вид")
                Spacer(Modifier.height(14.dp))
                Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
                    ThemeChip("Системная", "system", Modifier.weight(1f))
                    ThemeChip("Светлая", "light", Modifier.weight(1f))
                    ThemeChip("Тёмная", "dark", Modifier.weight(1f))
                }
            }
            Spacer(Modifier.height(18.dp))

            GlassCard {
                SectionLabel("Подключение")
                Spacer(Modifier.height(14.dp))
                Row(verticalAlignment = Alignment.CenterVertically) {
                    OrbButton(level = .18f, size = 88.dp)
                    Spacer(Modifier.width(16.dp))
                    Column(Modifier.weight(1f)) {
                        StatusPill(if (token.isBlank()) "Не подключено" else "Подключено", token.isNotBlank())
                        Spacer(Modifier.height(8.dp))
                        Text(if (token.isBlank()) "Добавьте mobile device token" else "Устройство готово к записи", color = Vd.ink, fontWeight = FontWeight.Bold)
                        Text("Последняя синхронизация: ${lastSyncText()}", color = Vd.muted, fontSize = 13.sp)
                    }
                }
            }
            Spacer(Modifier.height(18.dp))

            GlassCard {
                SectionLabel("О приложении")
                Spacer(Modifier.height(8.dp))
                Text("Personal Voice Task", color = Vd.ink, fontSize = 21.sp, fontWeight = FontWeight.Black)
                Text("Версия 0.1.0\nТокен: ${AppPrefs.maskedToken(this@SettingsActivity)}", color = Vd.muted, lineHeight = 21.sp)
            }
            Spacer(Modifier.height(18.dp))

            GlassCard {
                Text(
                    if (developerOpen) "Скрыть подключение" else "Для разработчиков",
                    color = Vd.ink,
                    fontSize = 18.sp,
                    fontWeight = FontWeight.Black,
                    modifier = Modifier
                        .fillMaxWidth()
                        .clip(RoundedCornerShape(16.dp))
                        .clickable { developerOpen = !developerOpen }
                        .padding(vertical = 6.dp)
                )
                if (developerOpen) {
                    Spacer(Modifier.height(12.dp))
                    CompactField("Backend URL", baseUrl, { baseUrl = it })
                    Spacer(Modifier.height(10.dp))
                    CompactField("Mobile device token", token, { token = it })
                    Spacer(Modifier.height(14.dp))
                    GradientButton("Сохранить подключение", modifier = Modifier.fillMaxWidth()) {
                        AppPrefs.save(this@SettingsActivity, baseUrl, token)
                        AppPrefs.markSynced(this@SettingsActivity)
                        Toast.makeText(this@SettingsActivity, "Настройки сохранены", Toast.LENGTH_SHORT).show()
                        load()
                    }
                }
            }
        }
    }

    @Composable
    private fun SettingSwitch(title: String, description: String, checked: Boolean, onChange: (Boolean) -> Unit) {
        Row(Modifier.fillMaxWidth().padding(top = 16.dp), verticalAlignment = Alignment.CenterVertically) {
            Column(Modifier.weight(1f)) {
                Text(title, color = Vd.ink, fontSize = 19.sp, fontWeight = FontWeight.Black)
                Text(description, color = Vd.muted, fontSize = 14.sp)
            }
            Switch(checked = checked, onCheckedChange = onChange)
        }
    }

    @Composable
    private fun ThemeChip(label: String, mode: String, modifier: Modifier) {
        val selected = theme == mode
        GlassCard(modifier.clickable {
            theme = mode
            AppPrefs.saveThemeMode(this@SettingsActivity, mode)
        }, radius = 18.dp) {
            Text(if (selected) "✓" else "○", color = if (selected) Vd.blue else Vd.muted, fontSize = 22.sp, fontWeight = FontWeight.Black)
            Text(label, color = if (selected) Vd.blue else Vd.ink, fontWeight = FontWeight.Bold)
        }
    }

    private fun lastSyncText(): String {
        val value = AppPrefs.lastSyncAt(this)
        if (value <= 0) return "ещё не было"
        return SimpleDateFormat("сегодня, HH:mm", Locale.getDefault()).format(Date(value))
    }
}
